### DISCUSSION TITLE
SumatraPDF does not respond to DDE commands

### DISCUSSION THREAD

**Discussion by georx** [sumatrapdfreader/sumatrapdf#2993] [Q&A]
In my C++ application (Windows 10) I want to open the PDF guide at a certain named-destination or page. So I implemented the DDE comunication with SumatraPDF (version 3.4.6 64-bit). The Open commands works fine using this syntax:

`[Open("<pdffilepath>"[,<newwindow>,<focus>,<forcerefresh>])]`

whereas the following commands to jump at a certain named-destination or page are not executed:

`[GotoNamedDest("<pdffilepath>","<destination name>")]`
`[GotoPage("<pdffilepath>",<page number>)]`

Please note that I created the named-destinations using Adobe Acrobat Pro.

What am I doing wrong?
Thanks in advance for any help.

**Comment by GitHubRulesOK** [comment]
exact sequence of call ??

SumatraPDF must be open and active as a service before second or sequential commands 
(may need short delay to ensure it is up and running) If the file is not found it will still open but show error
![image](https://user-images.githubusercontent.com/12754913/193354641-ce7f91d8-c8f0-44bf-9c71-efade0d83758.png)
The bad tab will stand if the correct second call is made
![image](https://user-images.githubusercontent.com/12754913/193355750-a3d54e9e-5ca6-496a-981d-4c594afa0184.png)


***destinations are highly specific per pdf document, so a sample of that exact type is needed for testing***
Here the named dests are Page2Line1 and here Page2Line2 (....continued)
![image](https://user-images.githubusercontent.com/12754913/193357544-07a9e788-aa5a-463e-b969-14e0a29b1be3.png)

A named destination like many other goto's needs to be zoomed or otherwise viewport scaled first to behave as expected

page number should work easiest so again a sample file and exact call sequence used is needed to say why your syntax does not.

**Comment by georx** [comment]
@GitHubRulesOK Thanks for your help.
Yes, the sequence of calls is correct and by the way `GotoNamedDest` and `GotoPage` are not responded even when the PDF file is already open in SumatraPDF. So the problem isn't the delay between DDE commands (please note that SumatraPDF DDE server is active because the DDE command `[Open("Sample.pdf", 0, 1, 1)]` does work).

Please find attached my file Sample.PDF, the named-destinations (create by Adobe Acrobat Pro) are the following:

`1_1_Come_leggere_questo_libro`
`Gli_esempi_di_lavori_concreti`
`Guida_concettuale_e_operativa`
`Le_indicazioni_su_come_fare`
`Marchi_di_altri_produttori` 

Once SumatraPDF is running, the sequence of my DDE calls are the following:

`[Open("Sample.pdf", 0, 1, 1)]` (As I said, this works)
`[GotoNamedDest("Sample.pdf", "Le_indicazioni_su_come_fare")]` (This call is ignored)

Can you please try these calls using the DDE commands tool you shown in your post above?

Where can I find that tool? I tried to download Chris Oldwood's one but it doesn't work in my 64 bit Windows (the program window appears and suddenly disappears). Is that for 32 bit Windows only?

Thanks again for your help.

[Sample.pdf](https://github.com/sumatrapdfreader/sumatrapdf/files/9690393/Sample.pdf)

**Reply by GitHubRulesOK** [reply]
At first glance the named destinations seem to be good, **however** note that **_"ideally"_** none should be more than 22 characters your shortest is "Marchi_di_altri_produttori" (25 chars) since there is an external web browsing limitation set by Adobe of 32 max including `nameddest=` thus perhaps related to any earlier webview issues.

I never have problems running DDEQuery.exe (v15r) on 64bit even in this sandbox, but it is a fickle tool as it is not always clear why a call did not run, after a while you get used to resetting the comms for single line tests,  Far simpler for bulk trials is cmd line ddecmd.exe (v16r) that can be scripted. Yes I tested it with 32bit SumatraPDF

The main point I note from your call (still to be tested) is filename is "loose" (sample.pdf is not unique) note that in my last example I included the exact file path and file name _`"<pdffilepath>"`_

**Reply by GitHubRulesOK** [reply]
Ok while testing your long names they failed but shortened to a couple of words they worked, however
![image](https://user-images.githubusercontent.com/12754913/193414762-fe2faa9b-5324-4c05-9538-c2be49ceea80.png)
On retesting the longest it also behaved, suggesting there is some issue over how they were formed and I see the source as MS Word
![image](https://user-images.githubusercontent.com/12754913/193418317-4a0a387b-73e5-49ff-9d9c-e699fd047c64.png)

**Reply by georx** [reply]
Thanks for your help.
Unfortunately in my code it still doesn't work.
I've been able to run DDEQuery.exe (v1.5) and I've also installed SumatraPDF 32 bit (v3.4.6) in order to reproduce your actions, but when I run DDEQuery.exe (having SumatraPDF already running) I cannot connect to SumatraPDF, the **Server | Connect** menu command only lists PROGMAN as available DDE server:

![image](https://user-images.githubusercontent.com/34890740/193418542-47a71547-2ec6-4ab4-a1b0-a7d7909137f6.png)

What am I doing wrong?

**Reply by GitHubRulesOK** [reply]
If detection is not working you can set the connection 
![image](https://user-images.githubusercontent.com/12754913/193420751-1319c4f2-042c-4df6-9e84-1fbb3f7c347c.png)
![image](https://user-images.githubusercontent.com/12754913/193420794-2234bf18-2f51-4b84-9378-784a91693f92.png)



I have edited the named dests thus they are not your source, which I am looking at now so yours were written by word so they look like this

```
<</Dests 80 0 R>>
endobj
75 0 obj
<</D(þÿ M a r c h i _ d i _ a l t r i _ p r o d u t t o r i)/S/GoTo>>
```
...
```
80 0 obj
<</Names[
(þÿ 1 _ 1 _ C o m e _ l e g g e r e _ q u e s t o _ l i b r o)81 0 R
(þÿ G l i _ e s e m p i _ d i _ l a v o r i _ c o n c r e t i)82 0 R
(þÿ G u i d a _ c o n c e t t u a l e _ e _ o p e r a t i v a)83 0 R
(þÿ L e _ i n d i c a z i o n i _ s u _ c o m e _ f a r e)84 0 R
(þÿ M a r c h i _ d i _ a l t r i _ p r o d u t t o r i)85 0 R
]>>
```
That is not Latin it is 16LE thus causes no matching, normaly a named dest is plain and simple Ansii text like a numeric [1345-33647-68897-56]

Working ones should look like
```
/Names [ 
(1_1_Come_leggere_questo_libro) 15 0 R 
(Gli_esempi_di_lavori_concreti) 16 0 R 
(Guida_concettuale_e_operativa) 17 0 R 
(Le_indicazioni_su_come_fare) 18 0 R 
(Marchi_di_altri) [ 14 0 R /XYZ 0 613.0421 null ]
(Marchi_di_altri_1) 19 0 R 
(Marchi_di_altri_12) [ 14 0 R /XYZ 0 357.25267 null ] 
(Marchi_di_altri_123) [ 14 0 R /XYZ 0 357.25267 null ] 
(Marchi_di_altri_1234) [ 14 0 R /XYZ 0 357.25267 null ] 
(Marchi_di_altri_12345) [ 14 0 R /XYZ 0 357.25267 null ]
(Marchi_di_altri_123456) [ 14 0 R /XYZ 0 357.25267 null ] 
(Marchi_di_altri_1234567) [ 14 0 R /XYZ 0 357.25267 null ] 
]
```

**Reply by georx** [reply]
I've solved !!

Other than the destination names lenght, the problem was that I was running my application from the C++ compiler and this wasn't executing "As Administrator". Running my application from its EXE "As Administrator" makes the `GotoNamedDest` DDE command working fine.

Many thanks for your help.

By the way, I'm still not able to send DDE commands to SumatraPDF by DDEQuery.exe. I've followed your suggestion to connect SUMATRA server, but when I enter the item and activate **Command | Execute** I get the following error:

![image](https://user-images.githubusercontent.com/34890740/193422499-864a887b-7383-428b-8b5d-7d5cc9ef89a0.png)

It isn't important now since I solved the communication in my application, but I would like to know what I'm doing wrong in case I need to use DDEQuery.exe again. Can you help me on this?

Thanks again.

**Reply by GitHubRulesOK** [reply]
those links are non conventional they are UTF-16 they should be plain text
